clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleZonalServerGroupCachingAgent.java

```diff
@@ -0,0 +1,906 @@
+        }
+      }
+    }
+    serverGroup.setStatefulPolicy(manager.getStatefulPolicy());
+    if (manager.getAutoHealingPolicies() != null && !manager.getAutoHealingPolicies().isEmpty()) {
+      serverGroup.setAutoHealingPolicy(manager.getAutoHealingPolicies().get(0));
+    }
+    populateAutoscaler(serverGroup, autoscaler);
+    return serverGroup;
+  }
+
+  private static void setRegionConfig(
+      GoogleServerGroup serverGroup,
+      InstanceGroupManager manager,
+      GoogleNamedAccountCredentials credentials) {
+
+    serverGroup.setRegional(manager.getZone() == null);
+
+    if (serverGroup.getRegional()) {
+      serverGroup.setRegion(Utils.getLocalName(manager.getRegion()));
+      List<String> zones = getZones(manager.getDistributionPolicy());
+      serverGroup.setZones(ImmutableSet.copyOf(zones));
+      serverGroup.setDistributionPolicy(new GoogleDistributionPolicy(zones));
+    } else {
+      String zone = Utils.getLocalName(manager.getZone());
+      serverGroup.setZone(zone);
+      serverGroup.setZones(ImmutableSet.of(zone));
+      serverGroup.setRegion(credentials.regionFromZone(zone));
+    }
+  }
+
+  private static List<String> getZones(DistributionPolicy distributionPolicy) {
+    if (distributionPolicy == null || distributionPolicy.getZones() == null) {
+      return ImmutableList.of();
+    }
+    return distributionPolicy.getZones().stream()
+        .map(z -> Utils.getLocalName(z.getZone()))
+        .collect(toImmutableList());
+  }
+
+  private static ImmutableMap<String, Integer> convertNamedPorts(InstanceGroupManager manager) {
+    if (manager.getNamedPorts() == null) {
+      return null;
+    }
+    return manager.getNamedPorts().stream()
+        .filter(namedPort -> namedPort.getName() != null)
+        .filter(namedPort -> namedPort.getPort() != null)
+        .collect(toImmutableMap(NamedPort::getName, NamedPort::getPort));
+  }
+
+  private static void setLaunchConfig(
+      GoogleServerGroup serverGroup,
+      InstanceGroupManager manager,
+      @Nullable InstanceTemplate instanceTemplate,
+      GoogleNamedAccountCredentials credentials,
+      ProviderCache providerCache) {
+
+    HashMap<String, Object> launchConfig = new HashMap<>();
+    launchConfig.put("createdTime", Utils.getTimeFromTimestamp(manager.getCreationTimestamp()));
+
+    if (instanceTemplate != null) {
+      List<AttachedDisk> disks = getDisks(instanceTemplate);
+      instanceTemplate.getProperties().setDisks(disks);
+      launchConfig.put("instanceTemplate", instanceTemplate);
+      launchConfig.put("launchConfigurationName", instanceTemplate.getName());
+      launchConfig.put("instanceType", instanceTemplate.getProperties().getMachineType());
+      launchConfig.put("minCpuPlatform", instanceTemplate.getProperties().getMinCpuPlatform());
+      setSourceImage(serverGroup, launchConfig, disks, credentials, providerCache);
+    }
+    serverGroup.setLaunchConfig(launchConfig);
```

Comment:
This looks like one of the few places we set a mutable collection on the `serverGroup`.

Reply:
Okay, fixed this and the other place (serverGroup.setAsg).

Sadly a GoogleServerGroup is going to be very difficult to make fully immutable because it has several fields that are mutable objects we don't control (from the Google Compute APIs) :/